From f61782bdfb82b99d88d0bddc696cdeb04a09b679 Mon Sep 17 00:00:00 2001
From: Yuhang Zhao <2546789017@qq.com>
Date: Sun, 19 Jun 2022 13:06:56 +0800
Subject: [PATCH] wangwenx190: Custom tweaks

Import changes list:

01. MSVC: Don't embed the debug symbols into binary files for release builds.
02. MSVC: Enable dead code elimination for release builds.
03. MSVC: Use new exception handling model.
04. MSVC: Enable whole program optimization & link time code generation for release shared builds.
05. MSVC: Enable control flow guard for release builds.
06. MSVC: Enable IntelJCC for release builds.
07. MSVC: Enable IntelCET for release builds.
08. Other minor tweaks.

Signed-off-by: Yuhang Zhao <2546789017@qq.com>
---
 scripts/toolchains/windows.cmake | 42 +++++++++++++++++++++++++-------
 1 file changed, 33 insertions(+), 9 deletions(-)

diff --git a/scripts/toolchains/windows.cmake b/scripts/toolchains/windows.cmake
index 7c8a54989..5a9a2290d 100644
--- a/scripts/toolchains/windows.cmake
+++ b/scripts/toolchains/windows.cmake
@@ -51,20 +51,44 @@ if(NOT _CMAKE_IN_TRY_COMPILE)
         set(CHARSET_FLAG)
     endif()
 
-    set(CMAKE_CXX_FLAGS " /nologo /DWIN32 /D_WINDOWS /W3 ${CHARSET_FLAG} /GR /EHsc /MP ${VCPKG_CXX_FLAGS}" CACHE STRING "")
-    set(CMAKE_C_FLAGS " /nologo /DWIN32 /D_WINDOWS /W3 ${CHARSET_FLAG} /MP ${VCPKG_C_FLAGS}" CACHE STRING "")
-    set(CMAKE_RC_FLAGS "-c65001 /DWIN32" CACHE STRING "")
+    set(COMMON_COMPILE_FLAGS "/nologo /DWIN32 /D_WINDOWS /W3 ${CHARSET_FLAG} /MP /bigobj /Zc:wchar_t")
+
+    set(CMAKE_CXX_FLAGS " /await:strict /EHsc /d2FH4 /GR /Zc:__cplusplus ${COMMON_COMPILE_FLAGS} ${VCPKG_CXX_FLAGS}" CACHE STRING "")
+    set(CMAKE_C_FLAGS " ${COMMON_COMPILE_FLAGS} ${VCPKG_C_FLAGS}" CACHE STRING "")
+    set(CMAKE_RC_FLAGS "/nologo /c65001 /DWIN32" CACHE STRING "")
 
     unset(CHARSET_FLAG)
 
-    set(CMAKE_CXX_FLAGS_DEBUG "/D_DEBUG ${VCPKG_CRT_LINK_FLAG_PREFIX}d /Z7 /Ob0 /Od /RTC1 ${VCPKG_CXX_FLAGS_DEBUG}" CACHE STRING "")
-    set(CMAKE_C_FLAGS_DEBUG "/D_DEBUG ${VCPKG_CRT_LINK_FLAG_PREFIX}d /Z7 /Ob0 /Od /RTC1 ${VCPKG_C_FLAGS_DEBUG}" CACHE STRING "")
-    set(CMAKE_CXX_FLAGS_RELEASE "${VCPKG_CRT_LINK_FLAG_PREFIX} /O2 /Oi /Gy /DNDEBUG /Z7 ${VCPKG_CXX_FLAGS_RELEASE}" CACHE STRING "")
-    set(CMAKE_C_FLAGS_RELEASE "${VCPKG_CRT_LINK_FLAG_PREFIX} /O2 /Oi /Gy /DNDEBUG /Z7 ${VCPKG_C_FLAGS_RELEASE}" CACHE STRING "")
+    set(GL_FLAG "/GL")
+    set(LTCG_FLAG "/LTCG")
+    if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+        set(GL_FLAG)
+        set(LTCG_FLAG)
+    endif()
+
+    set(COMMON_DEBUG_FLAGS "/D_DEBUG ${VCPKG_CRT_LINK_FLAG_PREFIX}d /Zi /Ob0 /Od /RTC1")
+
+    set(CMAKE_CXX_FLAGS_DEBUG "${COMMON_DEBUG_FLAGS} ${VCPKG_CXX_FLAGS_DEBUG}" CACHE STRING "")
+    set(CMAKE_C_FLAGS_DEBUG "${COMMON_DEBUG_FLAGS} ${VCPKG_C_FLAGS_DEBUG}" CACHE STRING "")
+
+    set(COMMON_RELEASE_FLAGS "${VCPKG_CRT_LINK_FLAG_PREFIX} /O2 /Ob3 /Oi /Oy /Ot /GF /GT /Gw /Gy /fp:fast /Zc:inline /DNDEBUG /Zi ${GL_FLAG} /QIntel-jcc-erratum /guard:cf")
+
+    set(CMAKE_CXX_FLAGS_RELEASE "${COMMON_RELEASE_FLAGS} ${VCPKG_CXX_FLAGS_RELEASE}" CACHE STRING "")
+    set(CMAKE_C_FLAGS_RELEASE "${COMMON_RELEASE_FLAGS} ${VCPKG_C_FLAGS_RELEASE}" CACHE STRING "")
 
     string(APPEND CMAKE_STATIC_LINKER_FLAGS_RELEASE_INIT " /nologo ")
-    set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "/nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF ${VCPKG_LINKER_FLAGS} ${VCPKG_LINKER_FLAGS_RELEASE}" CACHE STRING "")
-    set(CMAKE_EXE_LINKER_FLAGS_RELEASE "/nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF ${VCPKG_LINKER_FLAGS} ${VCPKG_LINKER_FLAGS_RELEASE}" CACHE STRING "")
+
+    set(COMMON_LINK_FLAGS "/nologo /CETCOMPAT /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF /OPT:LBR ${LTCG_FLAG} /GUARD:CF")
+
+    set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${COMMON_LINK_FLAGS} ${VCPKG_LINKER_FLAGS} ${VCPKG_LINKER_FLAGS_RELEASE}" CACHE STRING "")
+    set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${COMMON_LINK_FLAGS} ${VCPKG_LINKER_FLAGS} ${VCPKG_LINKER_FLAGS_RELEASE}" CACHE STRING "")
+
+    unset(GL_FLAG)
+    unset(LTCG_FLAG)
+    unset(COMMON_LINK_FLAGS)
+    unset(COMMON_RELEASE_FLAGS)
+    unset(COMMON_DEBUG_FLAGS)
+    unset(COMMON_COMPILE_FLAGS)
 
     string(APPEND CMAKE_STATIC_LINKER_FLAGS_DEBUG_INIT " /nologo ")
     string(APPEND CMAKE_SHARED_LINKER_FLAGS_DEBUG_INIT " /nologo ${VCPKG_LINKER_FLAGS} ${VCPKG_LINKER_FLAGS_DEBUG} ")
-- 
2.39.0.windows.1

